yt/yt/client/table_client/composite_compare.cpp

```diff
@@ -276,4 +314,119 @@ TFingerprint CompositeFarmHash(TYsonStringBuf value)
 
 ////////////////////////////////////////////////////////////////////////////////
 
+std::optional<TYsonString> TruncateCompositeValue(TYsonStringBuf value, i64 size)
+{
+    YT_VERIFY(value.GetType() == EYsonType::Node);
+
+    YT_VERIFY(size >= 0);
+    if (!size) {
+        return {};
+    }
+
+    TMemoryInput valueIn(value.AsStringBuf());
+    TYsonPullParser valueParser(&valueIn, EYsonType::Node);
+
+    TString truncatedYson;
+    TStringOutput output(truncatedYson);
+    output.Reserve(std::min(size, std::ssize(value.AsStringBuf())));
+    TCheckedInDebugYsonTokenWriter writer(&output);
+
+    i64 unclosedListCount = 0;
+    bool emptyResult = true;
+
+    for (auto remainingBytes = size;;) {
+        const auto item = valueParser.Next();
+
+        // Both braces of the outmost list are not considered to be inside an enclosing list.
+        bool isInsideList = unclosedListCount >= 2 || (unclosedListCount == 1 && item.GetType() != EYsonItemType::EndList);
```

Comment:
Can't we write `isInsideList = unclosedListCount >= 1`?

GetMinResultingSize() returns `0` in case of `EndList`.

Reply:
Yeah, I just wasn't sure how to express this parameter logically. It would probably be simpler that way + comment.